databricks_cli/stack/api.py

```diff
@@ -0,0 +1,350 @@
+
+    def _parse_config_file(self, filename):
+        """
+        Parse the json stack configuration template to a readable dict format.
+
+        :param filename: File path of the JSON stack configuration template.
+        :return: dict of parsed JSON stack config template.
+        """
+        stack_conf = {}
+        with open(filename, 'r') as f:
+            stack_conf = json.load(f)
+
+        return stack_conf
+
+    def _json_type_handler(self, obj):
+        """
+        Helper function to convert certain objects into a compatible JSON format.
+
+        Right now, converts a datetime object to an integer timestamp.
+
+        :param obj: Object that may be a datetime object.
+        :return: Timestamp integer if object is a datetime object.
+        """
+        if isinstance(obj, datetime):
+            # Get timestamp of datetime object- works with python2 and 3
+            return int(time.mktime(obj.timetuple()))
+        raise TypeError("Object of type '%s' is not JSON serializable" % type(obj))
+
+    def _generate_stack_status_path(self, stack_path):
+        """
+        Given a path to the stack configuration template JSON file, generates a path to where the
+        deployment status JSON will be stored after successful deployment of the stack.
+
+        :param stack_path: Path to the stack config template JSON file
+        :return: The path to the stack status file.
+
+        >>> self._generate_stack_status_path('./stack.json')
+        './stack.deployed.json'
+        """
+        stack_path_split = stack_path.split('.')
+        stack_path_split.insert(-1, STACK_STATUS_INSERT)
+        return '.'.join(stack_path_split)
+
+    def _load_stack_status(self, status_path):
+        """
+        Loads the deployment status metadata for a stack given a path to the stack configuration
+        template JSON file. Looks for the default local stack status path generated from
+        _generate_stack_status_path.
+
+        When loaded, the stack resource configurations from the past deployment will be loaded into
+        self.deployed_resource_config, using the RESOURCE_ID field of each resource as a key in the
+        dictionary.
+        The output from the databricks server of the deployment of each resource will also be
+        loaded in self.deployed_resources in the same way.
+
+        :param status_path: path to JSON stack configuration template.
+        :return: The dict of parsed JSON of the stack deployment status.
+        If path doesn't exist, will return an empty dict.
+        """
+        stack_status = {}
+        try:
+            if os.path.exists(status_path):
+                with open(status_path, 'r') as f:
+                    stack_status = json.load(f)
+                click.echo("Using deployment status file at %s" % status_path)
+
+                # Store map from resource_id to past deployment configuration of resources.
+                self.previous_deploy_resource_config_map = \
```

Comment:
After another pass, it seems that `previous_deploy_resource_config_map` and `previous_deploy_resource_status_map` are only needed in `deploy_stack` method. I am not sure why it is necessary to have those maps as global states -- using global variables always make dependency complex and are error prone. How about organize the deploy_stack method as follows:
(1) Read config as dict
(2) Read status as dict
(3) Validate config and status (you don't have to cover all cases, just obvious ones, we can always improve this later)
(4) Deploy resources one by one
helpers such as get_physical_id can be define with in the `deploy_stack` method. All the other methods in this class do not need to care about these corner cases, because `deploy_stack` have done all the dirty work.
Pros: reduce lots of catch-throw and validation codes, and thusly increase readabilities. Most of the methods can have minimal logics inside them because corner cases are taken care of.